### ISSUE TITLE
Request: Disable CSD on Wayland

### ISSUE THREAD

**Issue by 9ary** [lah7/gtk3-classic#28]
This patchset intentionally leaves CSD enabled on Wayland, because with GNOME, you can't really get away without it. However, on e.g. sway, you should be able to disable it safely.
[I've attempted to do that](https://github.com/Streetwalrus/gtk3-mushrooms/commit/bcc1f12b32e3c7cc2f7a361016897c1f1d2886a1), but it doesn't seem to work as intended.
File picker windows lose their shadow, and `swaymsg -t get_tree` does not report them as CSD windows anymore, but the window layout stays the same, with the CSD top bar instead of having the buttons at the bottom. On X11, CSD is still disabled properly.

It'd be nice to get better support for this use case.

**Comment by acutbal** [comment]
Yes, it would be nice also to get this feature  for kwin/Plasma.
:smile:

**Comment by RobinBoers** [comment]
I tried to disable CSD in my fork. I compiled and it seems to be working, but I think I broke some other things now, because the window controls are back in Xorg and Seahorse looks different. With the package from the AUR it looks like this:

![2021-10-11-175234_342x511_scrot](https://user-images.githubusercontent.com/60298132/136822637-3a2b7a3e-8721-4393-8c84-b3ab1a5ce4fd.png)

And with my version it looks like this:

![2021-10-11-175151_805x870_scrot](https://user-images.githubusercontent.com/60298132/136822653-3bab4c91-a53e-437c-b6b3-3a7523f3e542.png)

I have no idea how C works, and I just removed the if statement that said X11, so I have no clue why the window controls are back or why seahorse looks differently.

Edit: the file dialog mentioned in the first comment seems to behave how it is supposed to now.

Edit2: the file dialog works in all applications on my system exept the widget factory

**Comment by emmatebibyte** [comment]
This works to disable csd on plasma wayland for me:
https://github.com/ZaWertun/gtk3-nocsd

**Comment by rkunschke** [comment]
> This works to disable csd on plasma wayland for me: https://github.com/ZaWertun/gtk3-nocsd

gtk3-nocsd is not longer developed and has several problems.

**Comment by emmatebibyte** [comment]
> > This works to disable csd on plasma wayland for me: https://github.com/ZaWertun/gtk3-nocsd
> 
> gtk3-nocsd is not longer developed and has several problems.

right, but I was wondering/hoping gtk3-csd could lead to some insight if it hasn't already

**Comment by Ph42oN** [comment]
I would like to see this too, worst part is file chooser dialog where cancel and ok buttons are at the top instead of bottom like its on xorg or xwayland. I tried gtk3-nocsd, and while it does disable CSD, cancel and ok buttons stay at top. Currently only solution i found is to run gtk apps with GDK_BACKEND=x11.

**Comment by droc12345** [comment]
Not sure how far back it goes, but using gtk 3.24.43, with this patch, csd seems to be disabled on wayland
Use in conjunction with GTK_CSD=0, and telling the wayland compositor that you want ssd.
Use GTK_CSD=1 to turn on csd. 
Been running all  day and haven't noticed a problem.
Feel free to use it or add it to the collection.

```--- a/gtk/gtkwindow.c   2024-07-10 12:03:14.000000000 -0500
+++ b/gtk/gtkwindow.c   2025-01-25 08:13:59.361890146 -0600
@@ -6142,14 +6142,6 @@
     return TRUE;
 #endif

-#ifdef GDK_WINDOWING_WAYLAND
-  if (GDK_IS_WAYLAND_DISPLAY (gtk_widget_get_display (GTK_WIDGET (window))))
-    {
-      GdkDisplay *gdk_display = gtk_widget_get_display (GTK_WIDGET (window));
-      return !gdk_wayland_display_prefers_ssd (gdk_display);
-    }
-#endif
-
 #ifdef GDK_WINDOWING_WIN32
   if (g_strcmp0 (csd_env, "0") != 0 &&
       GDK_IS_WIN32_DISPLAY (gtk_widget_get_display (GTK_WIDGET (window))))```


Edit to add: I checked back to version 3.24.20 and this patch will still work, though for early versions it's
appx line 6117 instead of 6142, patch should handle the offset with fuzzing set.
Enjoy

**Comment by lah7** [comment]
Thanks @droc12345 for providing a patch. I've opened a draft pull request at #129 for testing. [There's a Arch pkg artifact for that build.](https://github.com/lah7/gtk3-classic/actions/runs/14535141410) In my limited testing, it only worked for Meld (3.20.4). GNOME Disks (3.38.2) and `gtk3-widget-factory` kept using CSDs, which does not under X11.

@Ph42oN's workaround suggestion to use `GDK_BACKEND=x11` just works. I'll add it to the README as a Wayland tip.

Not sure I'll merge #129 (patch above) yet, since it's not feature partial with X11 as I'd hoped. The ideal outcome would be to install `gtk3-classic` and CSDs go away, like X11. Unless, the majority here think it's good enough and doesn't cause any regressions with the GTK programs you use under Wayland.

Of course, if the GTK app is new enough these days, it probably uses the `libhandy` library, which sadly can't be patched with GTK alone (#32)
